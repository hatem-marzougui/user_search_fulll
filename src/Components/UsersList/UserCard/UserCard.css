.user-card {
  height: 300px;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 16px;
  border: 1px solid rgb(196, 196, 214);
  border-radius: 15px;
  background-color: #f9f9f9;
  position: relative;
  background-color: #d5d5d5;
  box-shadow: 0 2px 4px rgb(122, 116, 116);
}

.avatar {
  vertical-align: middle;
  width: 70px;
  height: 70px;
  border-radius: 50%;
}
.viewProfile_button {
  cursor: pointer;
  border-radius: 7px;
  outline: none;
  border: 1px solid #0a9aed;
  background-color: #00a1ff;
  height: 2.3rem;
  width: 55%;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
}
.checkbox {
  margin-right: 10px;
  cursor: pointer;
  width: 18px;
  height: 18px;
  position: absolute;
  top: 5px;
  left: 5px;
}
