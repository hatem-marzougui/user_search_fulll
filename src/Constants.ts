export const HEADER_TITLE: string = "Github Search";
export const VIEW_PROFILE_BUTTON_TEXT: string = "View Profile";
export const LOGIN: string = "Login";
export const EDIT_MODE: string = "Edit Mode";
export const DUPLICATE_USERS: string = "Duplicate Users";
export const SEARCH_PLACEHOLDER: string = "Search...";
export const USERS_API_ENDPOINT: string = "/search/users";
export const RATE_LIMIT_ERROR_MESSAGE: string =
  "GitHub request limit reached. Please try again in a minute.";
export const EMPTY_SEARCH_QUERY_ERROR: string = "Please enter a search query";
export const NO_RESULTS_FOUND: string = "No result found  ";
export const USER_CARD_AVATAR_ALT: string = "User Avatar";
export const USER_CARD_ID_LABEL: string = "ID:";
export const USER_CARD_LOGIN_LABEL: string = "Login:";
export const USER_CARD_VIEW_PROFILE_BUTTON_TEXT: string = "View Profile";
export const RATE_LIMIT: string = "rate limit";
export const ELEMENT_SELECTED: string = "elements selected";
