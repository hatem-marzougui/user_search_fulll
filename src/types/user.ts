export type User = {
  id: number;
  login: string;
  avatar_url: string;
};

export type GithubUserResponse = {
  total_count: number;
  incomplete_results: boolean;
  items: User[];
};
