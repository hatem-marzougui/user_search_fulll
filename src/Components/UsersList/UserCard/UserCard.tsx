import React from "react";
import "./UserCard.css";
import { LOGIN, VIEW_PROFILE_BUTTON_TEXT } from "../../../Constants";
import { User } from "../../../types/user";

type UserCardProps = {
  user: User;
  handleToggleSelect: (e: React.ChangeEvent<HTMLInputElement>) => void;
  isSelected: boolean;
  isEditMode: boolean;
};

const UserCard = ({
  user,
  handleToggleSelect,
  isSelected,
  isEditMode,
}: UserCardProps) => {
  return (
    <div className="user-card">
      {isEditMode && (
        <input
          type="checkbox"
          className="checkbox"
          onChange={handleToggleSelect}
          value={user.id}
          checked={isSelected}
        />
      )}
      <img src={user.avatar_url} alt="Avatar" className="avatar" />
      <h6>ID: {user.id}</h6>
      <h6>
        {LOGIN}: {user.login}
      </h6>
      <button className="viewProfile_button">{VIEW_PROFILE_BUTTON_TEXT}</button>
    </div>
  );
};

export default UserCard;
