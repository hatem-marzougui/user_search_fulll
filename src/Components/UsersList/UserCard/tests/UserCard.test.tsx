import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import UserCard from "../UserCard";
import { LOGIN, VIEW_PROFILE_BUTTON_TEXT } from "../../../../Constants";
import { User } from "../../../../types/user";

describe("UserCard", () => {
  const mockUser: User = {
    id: 123,
    login: "hatem",
    avatar_url: "https://avatars.githubusercontent.com/u/123456?v=4",
  };

  const handleToggleSelect = jest.fn();

  test("renders user info and button", () => {
    render(
      <UserCard
        user={mockUser}
        handleToggleSelect={handleToggleSelect}
        isSelected={false}
        isEditMode={false}
      />
    );

    expect(screen.getByText(`ID: ${mockUser.id}`)).toBeInTheDocument();
    expect(screen.getByText(`${LOGIN}: ${mockUser.login}`)).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: VIEW_PROFILE_BUTTON_TEXT })
    ).toBeInTheDocument();
    expect(screen.getByAltText("Avatar")).toHaveAttribute(
      "src",
      mockUser.avatar_url
    );
  });

  test("shows checkbox in edit mode and triggers handleToggleSelect", () => {
    render(
      <UserCard
        user={mockUser}
        handleToggleSelect={handleToggleSelect}
        isSelected={true}
        isEditMode={true}
      />
    );

    const checkbox = screen.getByRole("checkbox");
    expect(checkbox).toBeChecked();

    fireEvent.click(checkbox);
    expect(handleToggleSelect).toHaveBeenCalledTimes(1);
  });

  test("matches snapshot", () => {
    const { container } = render(
      <UserCard
        user={mockUser}
        handleToggleSelect={handleToggleSelect}
        isSelected={true}
        isEditMode={true}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
