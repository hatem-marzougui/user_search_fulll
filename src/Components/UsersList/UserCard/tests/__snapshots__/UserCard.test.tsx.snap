// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard matches snapshot 1`] = `
<div>
  <div
    class="user-card"
  >
    <input
      checked=""
      class="checkbox"
      type="checkbox"
      value="123"
    />
    <img
      alt="Avatar"
      class="avatar"
      src="https://avatars.githubusercontent.com/u/123456?v=4"
    />
    <h6>
      ID: 
      123
    </h6>
    <h6>
      Login
      : 
      hatem
    </h6>
    <button
      class="viewProfile_button"
    >
      View Profile
    </button>
  </div>
</div>
`;
